File: test-api.py
import urllib.request
import json
import uuid

import os
BASE = os.environ.get("BASE", "http://127.0.0.1:3001")

def api(method, path, data=None, token=None):
    url = f"{BASE}{path}"
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    
    body = json.dumps(data).encode() if data is not None else None
    req = urllib.request.Request(url, data=body, headers=headers, method=method)
    
    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        return json.loads(e.read())

passed = 0
total = 0

def test(name, result, expected_key=None):
    global passed, total
    total += 1
    if expected_key and expected_key in result and result.get("success", False):
        passed += 1
        print(f"  ✅ {name}")
    elif not expected_key and result:
        passed += 1
        print(f"  ✅ {name}")
    else:
        print(f"  ❌ {name}: {result}")

print("=" * 40)
print(" CopyCloud API Test Suite")
print("=" * 40)

# 1. Health
print("\n1. Health Check")
r = api("GET", "/health")
test("Health endpoint", r)

# 2. Register with unique email
email = f"test{uuid.uuid4().hex[:8]}@test.com"
print(f"\n2. Auth - Register ({email})")
r = api("POST", "/api/auth/register", {"email": email, "password": "pass123"})
test("Register new user", r, "token")
token = r.get("token", "")
user_id = r.get("user", {}).get("id", "")
print(f"     User: {user_id}")
print(f"     Token: {token[:30]}...")

# 3. Login
print("\n3. Auth - Login")
r = api("POST", "/api/auth/login", {"email": email, "password": "pass123"})
test("Login existing user", r, "token")
# Use the latest token from login
token = r.get("token", token)
print(f"     Token: {token[:30]}...")

# 4. Add Clipboard
print("\n4. Clipboard - Add")
r = api("POST", "/api/clipboard", {
    "content_type": "text",
    "encrypted_content": "Hello World!",
    "metadata": {"size": 12},
    "device_id": "device-1"
}, token)
test("Add clipboard item", r, "data")
clip_id = r.get("data", {}).get("id", "")
print(f"     Clip ID: {clip_id}")

# 5. Get Clipboard
print("\n5. Clipboard - Get All")
r = api("GET", "/api/clipboard", token=token)
test("Get clipboard items", r, "data")
count = len(r.get("data", []))
print(f"     Items: {count}")

# 6. Pin Clipboard
print("\n6. Clipboard - Pin")
r = api("PATCH", f"/api/clipboard/{clip_id}/pin", data={}, token=token)
test("Pin clipboard item", r, "data")
pinned = r.get("data", {}).get("metadata", {}).get("pinned", False)
print(f"     Pinned: {pinned}")

# 7. Delete Clipboard
print("\n7. Clipboard - Delete")
r = api("DELETE", f"/api/clipboard/{clip_id}", token=token)
test("Delete clipboard item", r)

# 8. Register Device
print("\n8. Device - Register")
r = api("POST", "/api/devices/register", {
    "name": "My PC",
    "platform": "windows"
}, token)
test("Register device", r, "data")
dev_id = r.get("data", {}).get("id", "")
print(f"     Device ID: {dev_id}")

# 9. Get Devices
print("\n9. Device - Get All")
r = api("GET", "/api/devices", token=token)
test("Get devices", r, "data")
print(f"     Devices: {len(r.get('data', []))}")

# 10. Update Device Status
print("\n10. Device - Update Status")
r = api("PATCH", f"/api/devices/{dev_id}/status", {"is_online": False}, token)
test("Update device status", r, "data")

# 11. Delete Device
print("\n11. Device - Delete")
r = api("DELETE", f"/api/devices/{dev_id}", token=token)
test("Delete device", r)

# Summary
print("\n" + "=" * 40)
print(f" Results: {passed}/{total} tests passed")
if passed == total:
    print(" ALL TESTS PASSED! ✅")
else:
    print(f" {total - passed} tests failed ❌")
print("=" * 40)

File: scripts/test-api.ts
import * as fs from 'fs';
import * as path from 'path';

interface TestResult {
  endpoint: string;
  method: string;
  status: number;
  success: boolean;
  duration: number;
  error?: string;
}

const BASE_URL = 'http://localhost:3737';
const results: TestResult[] = [];

async function testEndpoint(
  endpoint: string,
  method: string = 'GET',
  body?: any,
  headers?: Record<string, string>
): Promise<TestResult> {
  const start = Date.now();
  
  try {
    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
    };
    
    if (body) {
      options.body = JSON.stringify(body);
    }
    
    const response = await fetch(`${BASE_URL}${endpoint}`, options);
    const duration = Date.now() - start;
    
    return {
      endpoint,
      method,
      status: response.status,
      success: response.ok,
      duration,
    };
  } catch (error) {
    return {
      endpoint,
      method,
      status: 0,
      success: false,
      duration: Date.now() - start,
      error: error instanceof Error ? error.message : 'Unknown error',
    };
  }
}

async function runTests() {
  console.log('🧪 CopyCloud API Test Suite\n');
  console.log(`Testing: ${BASE_URL}\n`);
  
  // Health check
  console.log('1. Health Check');
  results.push(await testEndpoint('/health'));
  
  // API info
  console.log('2. API Info');
  results.push(await testEndpoint('/api'));
  
  // Auth endpoints
  console.log('3. Auth Endpoints');
  results.push(await testEndpoint('/api/auth/register', 'POST', {
    email: 'test@example.com',
    password: 'password123',
  }));
  results.push(await testEndpoint('/api/auth/login', 'POST', {
    email: 'test@example.com',
    password: 'password123',
  }));
  
  // Get token for authenticated endpoints
  let token = '';
  try {
    const loginResponse = await fetch(`${BASE_URL}/api/auth/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email: 'test@example.com', password: 'password123' }),
    });
    const loginData = await loginResponse.json();
    token = loginData.token || '';
  } catch {}
  
  const authHeaders = token ? { Authorization: `Bearer ${token}` } : {};
  
  // Clipboard endpoints
  console.log('4. Clipboard Endpoints');
  results.push(await testEndpoint('/api/clipboard', 'GET', undefined, authHeaders));
  results.push(await testEndpoint('/api/clipboard', 'POST', {
    content_type: 'text',
    preview: 'Test clip',
    metadata: { size: 9 },
  }, authHeaders));
  
  // Device endpoints
  console.log('5. Device Endpoints');
  results.push(await testEndpoint('/api/devices', 'GET', undefined, authHeaders));
  results.push(await testEndpoint('/api/devices/register', 'POST', {
    device_id: 'test-device-123',
    name: 'Test Device',
    platform: 'web',
  }, authHeaders));
  
  // Print results
  console.log('\n📊 Test Results:\n');
  
  const passed = results.filter(r => r.success).length;
  const failed = results.filter(r => !r.success).length;
  
  results.forEach((result, index) => {
    const status = result.success ? '✅' : '❌';
    const statusColor = result.success ? '\x1b[32m' : '\x1b[31m';
    const resetColor = '\x1b[0m';
    
    console.log(
      `${status} ${result.method.padEnd(6)} ${result.endpoint.padEnd(30)} ` +
      `${statusColor}${result.status}${resetColor} ${result.duration}ms` +
      (result.error ? ` (${result.error})` : '')
    );
  });
  
  console.log(`\n📈 Summary: ${passed} passed, ${failed} failed, ${results.length} total`);
  console.log(`⏱️  Total duration: ${results.reduce((sum, r) => sum + r.duration, 0)}ms`);
  
  // Save results
  const reportPath = path.join(__dirname, '../test-results.json');
  fs.writeFileSync(reportPath, JSON.stringify(results, null, 2));
  console.log(`\n📄 Results saved to: ${reportPath}`);
}

runTests().catch(console.error);
